File: coding-challenges/isBST.py
class BSTNode:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None


# def isBST(node):
#     return isBSTHelper(node, None, None)


# def isBSTHelper(node, maxi, mini):
#     if node is None:
#         return True
#     if maxi is not None and node.val >= maxi:
#         return False
#     if mini is not None and node.val <= mini:
#         return False
#     return isBSTHelper(node.left, node.val, mini) and isBSTHelper(node.right, maxi, node.val)

# inorder traverse
def isBST(node):
    res = []
    isBSTHelper(node, res)
    for i in range(0, len(res) - 1):
        if res[i] >= res[i+1]:
            return False
    return True


def isBSTHelper(node, arr):
    if node is not None:
        isBSTHelper(node.left, arr)
        arr.append(node.val)
        isBSTHelper(node.right, arr)


root = BSTNode(8)
root.left = BSTNode(7)
root.right = BSTNode(9)
root.left.left = BSTNode(6)
root.right.right = BSTNode(10)

print(isBST(root))

File: coding-challenges/isBST.js
class BSTNode {
    constructor(val) {
        this.val = val;
        this.left = null;
        this.right = null;
    }

    insertLeft(val) {
        this.left = new BSTNode(val);
        return this.left;
    }

    insertRight(val) {
        this.right = new BSTNode(val);
        return this.right;
    }
}

// inorder tranverse
// const isBST = (root) => {
//     const inOrder = (node, arr) => {
//         if (node) {
//             inOrder(node.left, arr);
//             arr.push(node.val);
//             inOrder(node.right, arr);
//         }
//     }
//     const res = [];
//     inOrder(root, res);
//     for (let i = 0; i < res.length - 1; i++) {
//         if (res[i + 1] <= res[i]) return false;
//     }
//     return true;
// }

// recursive
const isBST = (root) => {
    const bstHelper = (node, max, min) => {
        if (!node) return true;
        if (max && node.val >= max) return false;
        if (min && node.val <= min) return false;
        return bstHelper(node.left, node.val, min) && bstHelper(node.right, max, node.val);
    }
    return bstHelper(root, null, null);
}

const root = new BSTNode(10);
root.insertLeft(5);
root.insertRight(15);

console.log(isBST(root));